File: leetcode/algorithms/771_jewels_and_stones/main.java
import java.util.HashSet;
import java.util.Set;

class JewelsAndStones {
    public int numJewelsInStones(String jewels, String stones) {
        int[] numsArray = new int[128];
        for (char jewel : jewels.toCharArray()) {
            numsArray[jewel] = 1;
        }

        int count = 0;
        for (char stone : stones.toCharArray()) {
            count += numsArray[stone];
        }

        return count;
    }


    // Solution
    // Solution 1
    public int solution1(String jewels, String stones) {
        return stones.replaceAll("[^" + jewels + "]", "").length();
    }

    // Solution 2: HashSet
    public int solution2(String jewels, String stones) {
        Set<Character> jewelsSet = new HashSet<Character>();
        for (char jewel : jewels.toCharArray()) {
            jewelsSet.add(jewel);
        }

        int res = 0;
        for (char stone : stones.toCharArray()) {
            if (jewelsSet.contains(stone)) {
                res++;
            }
        }

        return res;
    }
}

File: leetcode/algorithms/771_jewels_and_stones/main.dart
class JewelsAndStones {
  int numJewelsInStones(String jewels, String stones) {
    List<int> numsArray = List<int>.filled(128, 0);
    for (int i = 0; i < jewels.length; i++) {
      numsArray[jewels[i].codeUnitAt(0)] = 1;
    }

    int count = 0;
    for (int i = 0; i < stones.length; i++) {
      count += numsArray[stones[i].codeUnitAt(0)];
    }

    return count;
  }


  // Solution
  // Solution 1
  int solution1(String jewels, String stones) {
    int ctr = 0;

    for (int i = 0; i < jewels.length; i++) {
      for (int j = 0; j < stones.length; j++) {
        if (jewels[i] == stones[j]) {
          ctr++;
        }
      }
    }

    return ctr;
  }

  // Solution 2
  int solution2(String jewels, String stones) {
    return stones.split("").where(jewels.contains).length;
  }
}
